runtime/libc/wasi_sdk_backing.c

```diff
@@ -297,14 +299,26 @@ wasi_errno_t wasi_snapshot_preview1_clock_time_get(
  * Read environment variable data.
  * The sizes of the buffers should match that returned by `environ_sizes_get`.
  * 
- * @param environ_retptr
- * @param environ_buf_retptr
+ * @param environ_baseretptr
+ * @param environ_buf_baseretptr
+ * @return WASI_ESUCCESS
  */
 wasi_errno_t wasi_snapshot_preview1_environ_get(
-    wasi_size_t environ_retptr, 
-    wasi_size_t environ_buf_retptr
+    wasi_size_t environ_baseretptr, 
+    wasi_size_t environ_buf_baseretptr
 ) {
-    wasi_unsupported_syscall(__func__);
+    wasi_size_t *environ_offset_vector = (wasi_size_t *)get_memory_ptr_for_runtime(environ_baseretptr, runtime_environc);
+    char *environ_buf = get_memory_ptr_for_runtime(environ_buf_baseretptr, runtime_environv_buf_len);
+    extern char **environ;
+
+    wasi_size_t environ_offset = 0;
+    for (wasi_size_t i = 0; i < runtime_environc; i++){
+        environ_offset_vector[i] = environ_buf_baseretptr + environ_offset;
+        strncpy(&environ_buf[environ_offset],environ[i], runtime_environv_buf_len - environ_offset);
```

Comment:
Why not just copy the whole vector at once?

Reply:
There's an extra bit of indirection. `environ` points to an array of pointers to strings, so they not guaranteed to consecutive.